File: 2023/ocaml/day02a/day02a.ml
open Printf

let total_red = 12
let total_green = 13
let total_blue = 14

let usage () =
  let progname = Sys.argv.(0) in
  printf "usage: %s <file>\n" progname;
  exit 1

let process contents =
  let result = ref 0 in
  String.split_on_char '\n' contents
  |> List.iter (fun line ->
      match String.split_on_char ':' line with
      | [game_str; draws_str] ->
        (match String.split_on_char ' ' game_str with
         | [_; game_num_str] ->
           let game_num = int_of_string game_num_str in
           let valid = ref true in
           String.split_on_char ';' draws_str
           |> List.iter (fun draw_str ->
              let trimmed_draw = String.sub draw_str 1 ((String.length draw_str) - 1) in
              String.split_on_char ',' trimmed_draw
              |> List.iter (fun color_amount ->
                let trimmed_color_amount =
                  if (String.get color_amount 0) == ' '
                    then (String.sub color_amount 1 ((String.length color_amount) - 1))
                    else color_amount
                in
                match String.split_on_char ' ' trimmed_color_amount with
                | [amount_str; color] ->
                  let amount = int_of_string amount_str in
                  (match color with
                  | "red" -> if amount > total_red then valid := false
                  | "green" -> if amount > total_green then valid := false
                  | "blue" -> if amount > total_blue then valid := false
                  | _ -> failwith "unknown color"
                  )
                | _ -> ()
              )
           );
           if !valid then result := !result + game_num
         | _ -> ()
        )
      | _ -> ()
    );
    !result

let () =
  if Array.length Sys.argv < 2 then
    usage ()
  else
    let filename = Sys.argv.(1) in
    let input_channel = open_in filename in
    try
      let contents = In_channel.input_all input_channel in
      let result = process contents in
        printf "result = %d\n" result
    with e ->
      raise e

File: 2023/c/day02a/day02a.c
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

#define TOTAL_RED 12
#define TOTAL_GREEN 13
#define TOTAL_BLUE 14

void usage(const char *progname) {
    printf("usage: %s <file>\n", progname);
    exit(EXIT_FAILURE);
}

void split_once(const char *source, char delim, char **first, char **second) {
    char *p = NULL;
    int pos = 0;

    if ( (p = strchr(source, delim)) != NULL ) {
        pos = p - source;
        if ( first != NULL ) {
            *first = strndup(source, pos);
        }
        if ( second != NULL ) {
            *second = strndup(source + pos + 1, strlen(source) - pos);
        }
    } else {
        if ( first != NULL ) {
            *first = NULL;
        }
        if ( second != NULL ) {
            *second = NULL;
        }
    }
}

uint32_t process(const char *filename) {
    uint32_t result = 0;
    char *line = NULL;
    FILE *infile = NULL;
    size_t len = 0;
    ssize_t nread = 0;
    char *game_part = NULL;
    char *draws_part = NULL;
    char *game_num_part = NULL;
    char *draw = NULL;
    char *rest = NULL;
    char *color_amount = NULL;
    char *color = NULL;
    char *amount_str = NULL;
    int game_num = 0, amount = 0;
    int last_draw = 0, last_color_amount = 0, valid = 0;

    if ( (infile = fopen(filename, "r")) == NULL ) {
        perror("fopen");
        exit(EXIT_FAILURE);
    }
    while ( (nread = getline(&line, &len, infile)) != -1 ) {
        len = strlen(line);
        line[len - 1] = '\0';
        split_once(line, ':', &game_part, &draws_part);
        split_once(game_part, ' ', NULL, &game_num_part);
        game_num = atoi(game_num_part);
        valid = 1;
        last_draw = 0;
        for ( ; ; ) {
            split_once(draws_part, ';', &draw, &rest);
            if ( draw != NULL ) {
                draws_part = rest;
            } else {
                draw = strdup(draws_part);
                last_draw = 1;
            }
            last_color_amount = 0;
            for ( ; ; ) {
                split_once(draw, ',', &color_amount, &rest);
                if ( color_amount != NULL ) {
                    draw = rest;
                } else {
                    color_amount = strdup(draw);
                    last_color_amount = 1;
                }
                split_once(color_amount + 1, ' ', &amount_str, &color);
                amount = atoi(amount_str);
                if ( strcmp(color, "red") == 0 && amount > TOTAL_RED ) {
                    valid = 0;
                } else if ( strcmp(color, "green") == 0 && amount > TOTAL_GREEN ) {
                    valid = 0;
                } else if ( strcmp(color, "blue") == 0 && amount > TOTAL_BLUE ) {
                    valid = 0;
                }
                if ( last_color_amount ) {
                    break;
                }
            }
            if ( last_draw ) {
                break;
            }
        }
        if ( valid ) {
            result += game_num;
        }
    }
    fclose(infile);
    return result;
}

int main(int argc, char *argv[]) {
    char *filename = NULL;
    uint32_t result = 0;

    if ( argc < 2 ) {
        usage(argv[0]);
    }
    filename = argv[1];
    result = process(filename);
    printf("result = %d\n", result);
    return EXIT_SUCCESS;
}
